modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpChannel.java

```diff
@@ -32,12 +32,12 @@ public class Netty4HttpChannel implements HttpChannel {
 
     @Override
     public void sendResponse(HttpResponse response, ActionListener<Void> listener) {
-        if (isOpen()) {
-            channel.writeAndFlush(response, Netty4TcpChannel.addPromise(listener, channel));
-        } else {
-            // No need to dispatch to the event loop just to fail this listener; moreover the channel might be closed because the whole
-            // node is shutting down, in which case the event loop might not exist any more so the channel promise cannot be completed.
-            listener.onFailure(new ClosedChannelException());
+        // We need to both guard against double resolving the listener and not resolving it in case of event loop shutdown so we need to
+        // use #notifyOnce here until https://github.com/netty/netty/issues/8007 is resolved.
+        var wrapped = ActionListener.notifyOnce(listener);
+        channel.writeAndFlush(response, Netty4TcpChannel.addPromise(wrapped, channel));
+        if (channel.eventLoop().isShutdown()) {
+            wrapped.onFailure(new TransportException("Cannot send HTTP response, event loop is shutting down."));
         }
```

Comment:
If the event loop stops after this check, I assume we can rely on Netty to release the response?

Reply:
Yeah shutting down the event loop involves draining the executor's queue. If we get here and the event loop _isn't_ shut down then we know the work was enqueued and thus will run.